src/dromajo_manycore.cpp

```diff
@@ -0,0 +1,314 @@
+/*
+ * API definitions for Dromajo to interact with the HammerBlade Manycore
+ */
+
+#include "dromajo_manycore.h"
+#include <iostream>
+
+// Helper functions
+bool get_fifo_empty(mc_fifo_t *fifo, bool _empty = false, uint32_t fifo_id = 0x0);
+uint32_t get_fifo_size(mc_fifo_t *fifo, uint32_t fifo_id = 0x0);
+bool get_fifo_full(mc_fifo_t *fifo, bool _full = false, uint32_t fifo_id = 0x0);
+void set_fifo_full(mc_fifo_t *fifo, uint32_t fifo_id);
+
+int get_fifo_credits(mc_fifo_t *fifo);
+bool fifo_write(mc_fifo_t *fifo, uint32_t fifo_id, uint32_t val);
+bool fifo_read(mc_fifo_t *fifo, uint32_t fifo_id, uint32_t *val);
+
+/*
+ * Initializes all FIFOs
+ * This function 
+ *    sets the full vector of the FIFO to all zeros (all FIFOs empty at the start)
+ *    defines the 4 32-bit FIFOs
+ *    sets the credits used to the minimum value
+ *    sets the init field to True
+ * @param[in] fifo - FIFO to initialize
+ */
+void mc_fifo_init(mc_fifo_t *fifo) {
+  for (int i = 0; i < 4; i++) {
+    fifo->fifo.push_back(std::queue<uint32_t>());
+    fifo->full.push_back(false);
+  }
+  fifo->credits = 0;
+  fifo->init = true;
+}
+
+/*
+ * Returns the state of the FIFO
+ * A 0 indicates the FIFO is empty
+ * A 1 indicates the FIFO is full
+ * @param[in] type - Type of FIFO to query
+ * @param[in] _full - Set to true if you need the fullness of a specific 32-bit FIFO
+ * @param[in] fifo_id - Chooses which 32-bit FIFO's fullness to return
+ * @returns FIFO full status
+ */
+bool mc_is_fifo_full(mc_fifo_type_t type, bool _full, uint32_t fifo_id) {
+  bool full= false;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
+      full = get_fifo_full(host_to_mc_req_fifo, _full, fifo_id);
+    break;
+    case FIFO_MC_TO_HOST_REQ:
+      full = get_fifo_full(mc_to_host_req_fifo, _full, fifo_id);
+    break;
+    case FIFO_MC_TO_HOST_RESP:
+      full = get_fifo_full(mc_to_host_resp_fifo, _full, fifo_id);
+    break;
+    default:
+    {
+      printf("Undefined/Unknown FIFO type!\nExiting...\n");
+      exit(-1);
+    }
+    break;
+  }
+  return full;
+}
+
+/*
+ * Returns if the FIFO is empty
+ * A 0 indicates the FIFO is not empty
+ * A 1 indicates the FIFO is empty
+ * @param[in] type - Type of FIFO to query
+ * @param[in] _full - Set to true if you need the fullness of a specific 32-bit FIFO
+ * @param[in] fifo_id - Chooses which 32-bit FIFO's fullness to return
+ * @returns FIFO empty status
+ */
+bool mc_is_fifo_empty(mc_fifo_type_t type, bool _empty, uint32_t fifo_id) {
+  bool empty = false;
+  switch (type) {
+    case FIFO_HOST_TO_MC_REQ:
```

Comment:
It doesn't seem like there's anything special about having different fifo types? Why not just have fifo_ids? That way we can expand this to more fifos in the future similar to the zynq-parrot setup

Reply:
I just kept it this way to make it easier to understand and mirror what we have in the manycore repository. Question is will we even need to expand to more FIFOs for the manycore? Or do you wanna make this setup more generic to handle other accelerators as well?

Reply:
Either ways the types are part of an enum and so effectively they are IDs :)

Reply:
I want to make it more generic to handle other accelerators. But yeah exactly it seems like there are two tags which do the exact same thing (the enum and the id)

Reply:
And in the manycore, here's a really simple use-case: multicore dromajo will require 3*N fifos. Could instantiate dromajo twice but not sure if that'll have a problem

Reply:
For the manycore: Isn't it more accurate to have > 1 Dromajo if we wanna emulate the multicore case?

For this PR: I guess you understand that the intention was to have a main type/ID which is the enum and each of the 32-bit FIFOs are identified by IDs. So what is the action item here? Should I just get rid of the 32-bit FIFO IDs and keep only the enum? So, writing to the FIFOs will look like - *addr = val. There could be a counter in the background that indicates which 32-bit FIFO the value goes to.

Reply:
>For the manycore: Isn't it more accurate to have > 1 Dromajo if we wanna emulate the multicore case?

No, dromajo is capable of multicore simulation. In fact, it is necessary to use multicore dromajo rather than several instances of single core dromajo to enable coherent memory between them

I would pass fifo pointers to each function that acts on them

Reply:
FIFO pointers are passed in the internal helper functions used to perform the actual function. The API functions are wrappers and require only a type to perform the function. The appropriate FIFO is chosen based on the switch..case statement.